pkg/statistics/handle/usage/indexusage/collector.go

```diff
@@ -0,0 +1,218 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package indexusage
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/pkg/parser/model"
+)
+
+// GlobalIndexID is the key type for indexUsageMap.
+type GlobalIndexID struct {
+	TableID int64
+	IndexID int64
+}
+
+// Sample is the data structure to store index usage information.
+type Sample struct {
+	LastUsedAt     time.Time
+	QueryTotal     uint64
+	KvReqTotal     uint64
+	RowAccessTotal uint64
+	// 0, 0-1, 1-10, 10-20, 20-50, 50-100, 100
+	PercentageAccess [7]uint64
+}
+
+var bucketBound = [6]float64{0, 0.01, 0.1, 0.2, 0.5, 1.0}
+
+func getIndexUsageAccessBucket(percentage float64) int {
+	if percentage == 0 {
+		return 0
+	}
+
+	bucket := 0
+	for i := 1; i < len(bucketBound); i++ {
+		if percentage >= bucketBound[i-1] && percentage < bucketBound[i] {
+			bucket = i
+			break
+		}
+	}
+	if percentage == 1.0 {
+		bucket = len(bucketBound)
+	}
+
+	return bucket
+}
+
+// NewSample creates a new data point for index usage.
+func NewSample(queryTotal uint64, kvReqTotal uint64, rowAccess uint64, tableTotalRows uint64) *Sample {
+	percentageAccess := [len(bucketBound) + 1]uint64{}
+
+	// if the `tableTotalRows` == 0, record the percentage as 1 and use the last bucket.
+	bucket := len(bucketBound)
+	if tableTotalRows > 0 {
+		rowAccessPercentage := float64(rowAccess) / float64(tableTotalRows)
+		bucket = getIndexUsageAccessBucket(rowAccessPercentage)
+	}
+	percentageAccess[bucket] = 1
+
+	return &Sample{
+		QueryTotal:       queryTotal,
+		KvReqTotal:       kvReqTotal,
+		RowAccessTotal:   rowAccess,
+		PercentageAccess: percentageAccess,
+		LastUsedAt:       time.Now(),
+	}
+}
+
+type indexUsage map[GlobalIndexID]Sample
+
+// SessionIndexUsageCollector is a list item that holds the index usage mapper. If you want to write or read mapper, you must lock it.
+// TODO: use a third-party thread-safe list implementation instead of maintaining the list manually.
+// TODO: merge this list into SessionStatsList.
+/*
+                            [session1]                [session2]                        [sessionN]
+                                |                         |                                 |
+                            update into              update into                       update into
+                                |                         |                                 |
+                                v                         v                                 v
+[StatsHandle.Head] --> [session1.IndexUsage] --> [session2.IndexUsage] --> ... --> [sessionN.IndexUsage]
```

Comment:
Can we rewrite this mechanism to use goroutines to collect information, which seems more safe and go-style?

Reply:
Done :heart: